### DISCUSSION TITLE
Metals command not available

### DISCUSSION THREAD

**Discussion by voodoojunk** [scalameta/nvim-metals#657] [Q&A] [answered]
Hi.

init.lua requires the scala module:

`require('scala')`

My `lua/scala.lua` looks like this:

```lua
vim.o.shortmess = string.gsub(vim.o.shortmess, 'F', '') .. 'c'

local ft = { "scala", "sbt", "java" }
local metals_config = require("metals").bare_config()

metals_config.settings = {
    showImplicitArguments = true
}
--- statusBarProvider
metals_config.init_options.statusBarProvider = "off"

-- make sure the correct capabilities for snippets are set
metals_config.capabilities = require("cmp_nvim_lsp").default_capabilities()

local nvim_metals_group = vim.api.nvim_create_augroup("nvim-metals", { clear = true })
vim.api.nvim_create_autocmd("FileType", {
pattern = ft,
callback = function()
  require("metals").initialize_or_attach(metals_config)
end,
group = nvim_metals_group,
})
```

The installation of nvim-metals was successful (I use Vim-Plug for package management .... yes, I know...)
 After the installation, I can see all the commands in the help page when running `:h MetalsCommands`, but the command are not actually accessible. For example `:MetalsInstall` returns:
 
```
 E492: Not an editor command: MetalsInstall
 ```
 
Trying to autocomplete commands Metals* doesn't show anything either....

I've also tried the absolute minimal config which yielded the same results:

```
vim.cmd([[autocmd FileType scala,sbt lua require("metals").initialize_or_attach({})]])
```

What am I missing?

**Comment by ckipp01** [comment]
Hey @voodoojunk, the error makes it seem like Metals isn't actually attached. When you see this can you try to do `:lua vim.print(vim.lsp.get_clients())`, and see if you have a client actually attached? If that table is empty, then they issue is with the autocmd.

**Reply by voodoojunk** [reply]
Hi. Thank you for your reply. 

My version of Nvim (0.9.4) does not have the `vim.lsp.get_clients` method (is that perhaps the cause of the issue?). I do however have the `vim.lsp.get_active_clients` and `vim.lsp.get_buf_clients` which both return an empty dict
You can look at my config above. It's almost a copy of the conf example in the repo. Can you spot anything wrong with it?
Also, can there be a client without an lsp server to attach to? (which I've attempted to install via the missing `:MetalsInstall` command)

**Reply by ckipp01** [reply]
> get_active_clients .... which both return an empty dict

Ah, then yes, this is for sure the issue. Metals isn't attached at all, so I'm assuming something isn't right with either the way you're either including the config or your autocommand. I'd recommend trying to create an autocmd that triggers on `FileType` like you have that just prints a message and see if the issue is there.

**Reply by voodoojunk** [reply]
OK. I've tried it and replaced the callback for the autocmd to print a message and it works just fine.
```
callback = function()
  --- require("metals").initialize_or_attach(metals_config)
  vim.print("HELLO THERE")
end,
```

Perhaps the problem is in the `metals_config`??? I've change the autocmd to print `metals_config`. Does this look OK? :
```
{
  capabilities = {
    textDocument = {
      completion = {
        completionItem = {
          commitCharactersSupport = true,
          deprecatedSupport = true,
          insertReplaceSupport = true,
          insertTextModeSupport = {
            valueSet = { 1, 2 }
          },
          labelDetailsSupport = true,
          preselectSupport = true,
          resolveSupport = {
            properties = { "documentation", "detail", "additionalTextEdits", "sortText", "filterText", "insertText", "textEdit", "insertTextFormat", "insertTextMode" }
          },
          snippetSupport = true,
          tagSupport = {
            valueSet = { 1 }
          }
        },
        completionList = {
          itemDefaults = { "commitCharacters", "editRange", "insertTextFormat", "insertTextMode", "data" }
        },
        contextSupport = true,
        dynamicRegistration = false,
        insertTextMode = 1
      }
    }
  },
  handlers = {},
  init_options = {
    compilerOptions = {},
    statusBarProvider = "off"
  },
  settings = {
    showImplicitArguments = true
  },
  tvp = {}
}
````

Is there anything else we can look at to debug the issue? Is there something wrong with my original callback?
Is there a way to debug/log what's happening during the attach?

**Reply by ckipp01** [reply]
> Perhaps the problem is in the metals_config??? I've change the autocmd to print metals_config. Does this look OK? :

Well you should be able to also just pass in `{}` here and still get all the commands. So we don't even need to worry about the on attach. One other thing _maybe_ you can look at is the nvim LSP logs. I'm not sure that will be helpful though since there is no attach, then there will be no logs. One other thing you can do is try with a standalone config file and directly pass that in when you start nvim just to ensure that works.

**Reply by voodoojunk** [reply]
Hi and thanks again.
 
lsp logs show nothing relevant.
Tried passing `{}` - didn't work.
One interesting thing is that the help shows `install_or_update()` but when I try to look at the modules members
```
:lua for k,v in pairs(require("metals")) do print(k, v) end
```
I do not see `install_or_update` although this is probably not the source of the issue, since I can't see any Metals* commands (Except in the help)

Like I said above, I even tried the bare minimum to debug:
```
vim.cmd([[autocmd FileType scala,sbt lua require("metals").initialize_or_attach({})]])
```

Any idea what I'm missing here? any other thoughts?

**Reply by voodoojunk** [reply]
FYI - even tried upgrading Nvim 0.9.4 -> 0.9.5 
didn't help...

**Reply by ckipp01** [reply]
Did you try with a standalone config file and now and passing it in directly to nvim when starting?

**Reply by voodoojunk** [reply]
Not sure what you mean so I tried this:

`custominit.lua`:
```lua
vim.cmd('source ~/.config/nvim/customplugins.vim')
require("metals").initialize_or_attach({})
```
`customplugings.vim`:
```
call plug#begin()
Plug 'preservim/nerdtree'
Plug 'neovim/nvim-lspconfig'
Plug 'hrsh7th/cmp-nvim-lsp'
Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/cmp-path'
Plug 'hrsh7th/cmp-cmdline'
Plug 'hrsh7th/nvim-cmp'
Plug 'hrsh7th/cmp-vsnip'
Plug 'hrsh7th/vim-vsnip'
Plug 'nvim-telescope/telescope.nvim'
Plug 'vim-airline/vim-airline'
Plug 'vim-airline/vim-airline-themes'
Plug 'ryanoasis/vim-devicons'

" --- metals
Plug 'nvim-lua/plenary.nvim'
Plug 'j-hui/fidget.nvim'
Plug 'scalameta/nvim-metals'

call plug#end()
```
 and then started nvim with `vim -u custominit.lua`.  No bueno....

Also tried creating a `custominit.vim`:
```
call plug#begin()
Plug 'preservim/nerdtree'
Plug 'neovim/nvim-lspconfig'
Plug 'hrsh7th/cmp-nvim-lsp'
Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/cmp-path'
Plug 'hrsh7th/cmp-cmdline'
Plug 'hrsh7th/nvim-cmp'
Plug 'hrsh7th/cmp-vsnip'
Plug 'hrsh7th/vim-vsnip'
Plug 'nvim-telescope/telescope.nvim'
Plug 'vim-airline/vim-airline'
Plug 'vim-airline/vim-airline-themes'
Plug 'ryanoasis/vim-devicons'

" --- metals
Plug 'nvim-lua/plenary.nvim'
Plug 'j-hui/fidget.nvim'
Plug 'scalameta/nvim-metals'

call plug#end()

autocmd FileType scala,sbt lua require("metals").initialize_or_attach({})
```
then started, Nvim with `vim -u custominit.vim`. Also no luck...

**Reply by ckipp01** [reply]
Yea, what I meant was using `-u configFile.lua` or something, but it seems that also didn't work. Honestly, I have no idea. I haven't seen something like this reported before and without having any sort of info or logs or more details about what is happening, then I'm not sure. One other place to possibly look would be in your cache dir. nvim-metals should have a log file there that if it is getting started at all, it should log. For example on mac mine is located at `~/.cache/nvim/nvim-metals/nvim-metals.log`

**Reply by voodoojunk** [reply, marked as answer]
Well ... no such luck. The log file doesn't exist (though the `~/.cache/nvim/nvim-metals` dir does)
I guess I'll have to try to install the language server manually and start it via nvim-lspconfig
I'm not a scala developer. I just need to debug some code once in a while (mostly Apache Spark) so I just need the basic stuff like jump to definition, etc...
Hopefully working with lspconfig would suffice for that 

Thanks again for all your help :-)

**Reply by voodoojunk** [reply]
Update - yeah.... installed metals server manually and using nvim-lspconfig to start it - works fine...
Thanks again...
